File: Cpp/Easy/BestTimeToBuyAndSellStock.cpp
/*****************************************************
 *
 *  121. Best Time to Buy and Sell Stock solution c++
 *
 *  Joel M. Gonzalez Rodriguez
 *
*/****************************************************

int maxProfit(vector<int> &prices)
{
    int l, r, profit, high;
    l = profit = high = 0;
    r = 1;

    while (r < prices.size())
    {
        if (prices[l] < prices[r])
        {
            profit = prices[r] - prices[l];
            high = max(high, profit);
        }

        else
            l = r;
        r++;
    }

    return high;
}

File: Python/Easy/Best_Time_to_Buy_and_Sell_Stock.py
'''

    121. Best Time to Buy and Sell Stock

    Solution

'''

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l = 0
        r = 1
        max_profit = 0

        while r < len(prices):
            if  prices[r] - prices[l] > max_profit:
                max_profit = prices[r] - prices[l]
            
            if prices[l] > prices[r]:
                l = r
                r += 1
            
            else:
                r += 1

        return max_profit
